### ISSUE TITLE
Flutter web - Parent widget is not scrolling when mouse over is on HtmlElementView Widget for iframe.

### ISSUE THREAD

**Issue by pushparajakln** [flutter/flutter#86021]
In Flutter web project, We are embedding Webview inside a List through HtmlElementView widget. Parent widget is not scrolling when the mouse over is on HtmlElementView Widget. We are getting height of web content and assigning that height to the container height. so, there is no scroll since content height and container height is same. but parent widget scroll is not blocking. The cursor doesn't move at all when it is inside HtmlElementView.  Please check this issue. Below are my code snippet,
```
import 'dart:html' as html;
import 'dart:ui' as ui;

html.window.addEventListener('message', listen);
iframeElement.id = 'extra-frame';
iframeElement.src = webViewModel.guideUrl;
iframeElement.style.border = 'none';

// ignore: undefined_prefixed_name
ui.platformViewRegistry.registerViewFactory(
iframeElement.id,
(int viewId) {
return iframeElement;
},
);

Container(
height: webViewModel.webViewHeight != null ? webViewModel.webViewHeight : 300,
child: HtmlElementView(
key: UniqueKey(),
viewType: iframeElement.id,
)
)
```

**Comment by darshankawar** [comment]
@pushparajakln 
There's a slightly different HTML markup for web app that was a breaking change. If you are writing your own custom HtmlElementView widgets, you'll need to change the markup of your custom view.

See this document for more details and reference and see if it helps: https://flutter.dev/docs/release/breaking-changes/platform-views-using-html-slots-web

**Comment by pushparajakln** [comment]
> @pushparajakln
> There's a slightly different HTML markup for web app that was a breaking change. If you are writing your own custom HtmlElementView widgets, you'll need to change the markup of your custom view.
> 
> See this document for more details and reference and see if it helps: https://flutter.dev/docs/release/breaking-changes/platform-views-using-html-slots-web

I am not facing issue for displaying web widget. I have mentioned the actual issue of scrolling. will it resolve ?. please check my question. I want to block scroll of HtmlElementView.

**Comment by darshankawar** [comment]
@pushparajakln 
Provide `flutter doctor -v` and a [complete minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) that we can use to verify this.

**Comment by pushparajakln** [comment]
Apples-MacBook-Pro:flutter pushparaj$ flutter doctor -v
[✓] Flutter (Channel dev, 2.3.0-16.0.pre, on macOS 11.0.1 20B50 darwin-x64, locale en-IN)
    • Flutter version 2.3.0-16.0.pre at /Applications/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fa5883b78e (7 weeks ago), 2021-05-21 13:04:03 -0700
    • Engine revision 2f067fc4c5
    • Dart version 2.14.0 (build 2.14.0-136.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/pushparaj/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (2 available)
    • Raja (mobile) • ba1368607aec5f90f5ad398205b2d228016a2102 • ios            • iOS 14.0.1
    • Chrome (web)  • chrome                                   • web-javascript • Google Chrome 91.0.4472.114

• No issues found!

Code:
```
class WebViewGuide extends StatefulWidget {
  final WebViewModel webViewModel;
  final bool isKey;
  WebViewGuide({this.webViewModel, this.isKey = true});

  @override
  _WebViewGuideState createState() => new _WebViewGuideState();
}

class WebViewModel {
  double webViewHeight;
  final String guideUrl;
  WebViewModel({this.webViewHeight, this.guideUrl});
}

class _WebViewGuideState extends State<WebViewGuide> {
  WebViewModel webViewModel;
  InAppWebViewController _webViewController;
  final html.IFrameElement iframeElement = html.IFrameElement();
  void listen(html.Event event) {
    var data = (event as html.MessageEvent).data;
    if (data['type'] != null && data['type'] == "height") {
      if (webViewModel.webViewHeight == 0 || webViewModel.webViewHeight == null) {
        setState(() {
          webViewModel.webViewHeight =
          data['value'] is String ? double.parse(data['value']) : data['value']
              .toDouble();
        });
      }
    }
  }
  @override
  void initState() {
    super.initState();
    webViewModel = widget.webViewModel;
    html.window.addEventListener('message', listen);
    iframeElement.id = 'extra-frame';
    iframeElement.src = webViewModel.guideUrl;
    iframeElement.style.border = 'none';

    // ignore: undefined_prefixed_name
    ui.platformViewRegistry.registerViewFactory(
      iframeElement.id,
          (int viewId) {
        return iframeElement;
      },
    );

  }

  @override
  void dispose() {
    _webViewController = null;
    html.window.removeEventListener('message', listen);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: webViewModel.webViewHeight != null ? webViewModel.webViewHeight : 300,
      child:
      HtmlElementView(
        key: UniqueKey(),
        viewType: iframeElement.id,
      )
      ,
    );
  }
}
```

Our JS code to post height to flutter app,
```
window.onload = (event) => {
    console.log('page is fully loaded');
    var documentHeight = document.body.scrollHeight;
    console.log(documentHeight)
    window.parent.postMessage({type:"height", value:documentHeight});
        document.body.addEventListener('click', event => {
        event.preventDefault();
            if($(event.target).closest('a').attr('href')!== undefined){
                var anchorHref = $(event.target).closest('a').attr('href');
                window.parent.postMessage({type:"link", value:anchorHref});
            }
    })
}
```

**Comment by darshankawar** [comment]
You seem to be using 3rd party plugin `InAppWebview` in your code:

`InAppWebViewController _webViewController;`

Flutter's official webview plugin is https://pub.dev/packages/webview_flutter
Try with this plugin and see if the behavior is still the same. If so, provide updated code sample using `webview` plugin.

**Comment by pushparajakln** [comment]
> You seem to be using 3rd party plugin `InAppWebview` in your code:
> 
> `InAppWebViewController _webViewController;`
> 
> Flutter's official webview plugin is https://pub.dev/packages/webview_flutter
> Try with this plugin and see if the behavior is still the same. If so, provide updated code sample using `webview` plugin.

You can ignore InAppWebViewController _webViewController; by commenting it.
```

class WebViewGuide extends StatefulWidget {
  final WebViewModel webViewModel;
  final bool isKey;
  WebViewGuide({this.webViewModel, this.isKey = true});

  @override
  _WebViewGuideState createState() => new _WebViewGuideState();
}

class WebViewModel {
  double webViewHeight;
  final String guideUrl;
  WebViewModel({this.webViewHeight, this.guideUrl});
}

class _WebViewGuideState extends State<WebViewGuide> {
  WebViewModel webViewModel;
  final html.IFrameElement iframeElement = html.IFrameElement();
  void listen(html.Event event) {
    var data = (event as html.MessageEvent).data;
    if (data['type'] != null && data['type'] == "height") {
      if (webViewModel.webViewHeight == 0 || webViewModel.webViewHeight == null) {
        setState(() {
          webViewModel.webViewHeight =
          data['value'] is String ? double.parse(data['value']) : data['value']
              .toDouble();
        });
      }
    }
  }
  @override
  void initState() {
    super.initState();
    webViewModel = widget.webViewModel;
    html.window.addEventListener('message', listen);
    iframeElement.id = 'extra-frame';
    iframeElement.src = webViewModel.guideUrl;
    iframeElement.style.border = 'none';

    // ignore: undefined_prefixed_name
    ui.platformViewRegistry.registerViewFactory(
      iframeElement.id,
          (int viewId) {
        return iframeElement;
      },
    );

  }

  @override
  void dispose() {
    _webViewController = null;
    html.window.removeEventListener('message', listen);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: webViewModel.webViewHeight != null ? webViewModel.webViewHeight : 300,
      child:
      HtmlElementView(
        key: UniqueKey(),
        viewType: iframeElement.id,
      )
      ,
    );
  }
}
```

**Comment by darshankawar** [comment]
@pushparajakln 
I tried coming up with a complete running example from your sample by updating it as required, but what `webviewmodel` to pass below ?

```
void main() {
  runApp(MaterialApp(
    home: WebViewGuide(webViewModel: ,isKey: true),
  ));
}
```

Hence, we were requesting a complete minimal sample.

**Comment by pushparajakln** [comment]
> @pushparajakln
> I tried coming up with a complete running example from your sample by updating it as required, but what `webviewmodel` to pass below ?
> 
> ```
I > void main() {
>   runApp(MaterialApp(
>     home: WebViewGuide(webViewModel: ,isKey: true),
>   ));
> }
> ```
> 
> Hence, we were requesting a complete minimal sample.

Please check my above sample code,  Class is there
 class WebViewModel {
  double webViewHeight;
  final String guideUrl;
  WebViewModel({this.webViewHeight, this.guideUrl});
}

Please do as like below and don't forget to send height from your html, place the WebViewGuide widget inside list view. 
runApp(MaterialApp(
    home: WebViewGuide(webViewModel: WebViewModel(guideUrl: <your html url>),isKey: true),
  ));

**Comment by darshankawar** [comment]
Thanks for the details. 

<details>
<summary>code sample</summary>

```

import 'dart:async';

import 'package:flutter/material.dart';
import 'dart:html' as html;

import 'package:webview_flutter/webview_flutter.dart';

void main() {
  runApp(MaterialApp(
    home: WebViewGuide(webViewModel: null,),
  ));
}


class WebViewGuide extends StatefulWidget {
  final WebViewModel webViewModel;
  final bool isKey;
  WebViewGuide({required this.webViewModel, this.isKey = true});

  @override
  _WebViewGuideState createState() => new _WebViewGuideState();
}

class WebViewModel {
  double webViewHeight;
  final String guideUrl;
  WebViewModel({required this.webViewHeight, required this.guideUrl});
}

class _WebViewGuideState extends State<WebViewGuide> {
  late WebViewModel webViewModel;
  final html.IFrameElement iframeElement = html.IFrameElement();

  late final Completer<WebViewController> _webViewController =
  Completer<WebViewController>();
  void listen(html.Event event) {
    var data = (event as html.MessageEvent).data;
    if (data['type'] != null && data['type'] == "height") {
      if (webViewModel.webViewHeight == 0 || webViewModel.webViewHeight == null) {
        setState(() {
          webViewModel.webViewHeight =
          data['value'] is String ? double.parse(data['value']) : data['value']
              .toDouble();
        });
      }
    }
  }
  @override
  void initState() {
    super.initState();
    webViewModel = widget.webViewModel;
    html.window.addEventListener('message', listen);
    iframeElement.id = 'extra-frame';
    iframeElement.src = webViewModel.guideUrl;
    iframeElement.style.border = 'none';

    // ignore: undefined_prefixed_name
    ui.platformViewRegistry.registerViewFactory(
      iframeElement.id,
          (int viewId) {
        return iframeElement;
      },
    );

  }

  @override
  void dispose() {
  //  _webViewController = null;
    html.window.removeEventListener('message', listen);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: webViewModel.webViewHeight != null ? webViewModel.webViewHeight : 300,
      child:
      HtmlElementView(
        key: UniqueKey(),
        viewType: iframeElement.id,
      )
      ,
    );
  }
}




```

</details>

<details>
<summary> flutter doctor -v </summary>

```
[✓] Flutter (Channel stable, 2.2.3, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.2.3 at /Users/dhs/documents/fluttersdk/flutter
    • Framework revision f4abaa0735 (4 days ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (4 available)
    • SM G975F (mobile) • RZ8M802WY0X • android-arm64 • Android 10 (API 29)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 91.0.4472.114

• No issues found!





```

</details>

**Comment by ditman** [comment]
I've been investigating this a little bit, and the current problem (which is a web browser limitation, not limited to flutter), is that when the user is scrolling on an iframe, the iframe "swallows" all the events.

You can see it here: https://jsfiddle.net/ditman/ymouk45a/ (open the JS debug console and see the difference between scrolling on top of the orange background vs scrolling on top of the iframe).

Since flutter (framework) is synthesizing all the scroll events, instead of using native scrollable elements, it never sees when the iframe finishes scrolling (in the example above, you can see that the parent element starts emitting "scroll" events, even when he mouse is on top of the iframe!)

The example above gives me ideas for a workaround in this case, and maybe a more permanent solution (which might be a Good New Package)...

1. "Workaround:" have you tried making the iframe have `pointer-events: none`? This only works If the user doesn't need to interact with the contents of the iframe, but it might be enough for your use-case (which I don't know).
2. A better solution would be to implement something like the #scrolling div in the example above, and listen to its scroll events (horizontal/vertical) to see if an iframe on top of it is being scrolled. This might result in very ugly JS/CSS, but it seems doable! (In fact, I can't believe this hasn't been attempted yet)
3. An even better solution would be to ensure that the iframe is cooperating with the framework, internally monitoring its mousewheel/scroll events, and broadcasting those OUT through postMessage to a listener setup in your Flutter app. Then, it'd be a matter of funneling those events into the engine so it can take care of scrolling once the iframe is done scrolling in any direction.

If we were to implement a fix for this, I think I'd start with 3 (introducing a more-or-less generic of passing messages to the flutter engine from postMessage, so we can have for example 2 flutter apps talking to each other through iframes), then option 2 would be slightly easier to build for the iframes we don't fully control (and 1 wouldn't be required at all).

(However, the scope of this change is large, as you've probably realized by the length of my explanation :P)

**Comment by n00bvn** [comment]
I'm encountering the same problem and still can't find a solution for this.

I tried to postMessage from iframe when it scrolls, but the tricky part here is:
1. I still need user interactions inside the iframe, so I do set the height of HtmlElementView to be the same as the height of iframe's document.
2. Since the iframe has full height, so the iframe's document does not response to onscroll event at all. At the same time, the HtmlElementView also does not response to scrolling at all.

So is there anyway to manually fix this? I'm wondering something like covering the HtmlElementView with custom scroll and delegate it's scrolling to parent widget, so we can both scroll the parent widget and allow other interactions like click inside the iframe. Just a thought...

I'd be really appreciated any help. I really need to make this work.

PS: this works totally ok on mobile with WebView.

**Comment by alexdess** [comment]
Hello,

I have exactly the same problem, I created a smart form in Flutter that I integrated in Wordpress in Iframe and the scroll does not work at all as explained by ditman. Does anyone have a temporary workaround while waiting for a Flutter update to fix this problem? It's a real problem for my site and I'm even hesitating to rewrite the form in angular Dart.
Thanks in advance.

**Comment by ditman** [comment]
Possibly duplicate of: https://github.com/flutter/flutter/issues/81081
